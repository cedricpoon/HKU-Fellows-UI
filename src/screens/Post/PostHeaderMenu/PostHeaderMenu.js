import React, { Component } from 'react';
import { Linking, Alert, Clipboard } from 'react-native';
import { Button, Icon, Text, Separator } from 'native-base';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { PopupMenu } from 'hkufui/components';

import { localize } from 'hkufui/locale';
import themeStyles from 'hkufui/theme/Styles';
import { email } from 'hkufui/config';
const locale = localize({ language: 'en', country: 'hk' });

import { onVote, onNotify, onAccept } from '../viewActions';
import styles from './Styles';

export class PostHeaderMenu extends Component {
  constructor(props) {
    super(props);

    this._voteUp = this._voteUp.bind(this);
    this._voteDown = this._voteDown.bind(this);
    this._enableNotification = this._enableNotification.bind(this);
    this._acceptAnswer = this._acceptAnswer.bind(this);
    this._reportAbuse = this._reportAbuse.bind(this);
  }

  _voteUp() {
    const { onVote, postId, topicId } = this.props;
    onVote({ postId, topicId, value: +1 });
  }

  _voteDown() {
    const { onVote, postId, topicId } = this.props;
    onVote({ postId, topicId, value: -1 });
  }

  _enableNotification() {
    const { onNotify, topicId } = this.props;
    onNotify({ topicId });
  }

  _reportAbuse() {
    const { address, subject, template, ref } = email;
    const { postId } = this.props;
    // open mail app to report
    Linking.openURL(`mailto:${address}?subject=${subject}&body=${template(postId)}`)
      .catch(() => {
        Alert.alert(
          locale['alert.noEmailTitle'],
          locale['alert.noEmailContent'](address, ref(postId)),
          [
            {text: locale['alert.copy'], onPress: () => Clipboard.setString(`${address} - ${ref(postId)}`)},
            {text: locale['toast.dismiss']}
          ]
        );
      });
  }

  _acceptAnswer() {
    const { onAccept, postId, topicId } = this.props;
    onAccept({ postId, topicId });
  }

  render() {
    const { onRef, index, solved, native, ...restProps } = this.props;

    return(
      <PopupMenu
        ref={onRef}
        { ...restProps } /* Proptypes handling on <PopupMenu /> */
      >
        {native && (
          <Button transparent iconLeft light onPress={this._voteUp}>
            <Icon name="thumb-up" type="MaterialCommunityIcons" style={themeStyles.icon}></Icon>
            <Text>{locale['header.thumbUp'](index)}</Text>
          </Button>
        )}
        {native && (
          <Button transparent iconLeft light onPress={this._voteDown}>
            <Icon name="thumb-down" type="MaterialCommunityIcons" style={themeStyles.icon}></Icon>
            <Text>{locale['header.thumbDown'](index)}</Text>
          </Button>
        )}
        {native && (
          <Button transparent iconLeft warning disabled={solved} onPress={this._acceptAnswer}>
            <Icon name="checkbox-marked-circle-outline" type="MaterialCommunityIcons" style={themeStyles.icon}></Icon>
            <Text>{locale['header.solved']}</Text>
          </Button>
        )}
        {native && (
          <Separator style={styles.separator} />
        )}
        {native && (
          <Button transparent iconLeft light onPress={this._enableNotification}>
            <Icon name="notifications" type="MaterialIcons" style={themeStyles.icon}></Icon>
            <Text>{locale['header.notifications']}</Text>
          </Button>
        )}
        <Button transparent iconLeft danger onPress={this._reportAbuse}>
          <Icon name="report" type="MaterialIcons" style={themeStyles.icon}></Icon>
          <Text>{locale['header.abuse']}</Text>
        </Button>
      </PopupMenu>
    );
  }
}

PostHeaderMenu.propTypes = {
  postId: PropTypes.string.isRequired,
  topicId: PropTypes.string.isRequired,
  onRef: PropTypes.func,
  native: PropTypes.bool,
  index: PropTypes.number.isRequired,
  solved: PropTypes.bool,
  onVote: PropTypes.func,
  onAccept: PropTypes.func,
  onNotify: PropTypes.func
}

const mapDispatchToProps = dispatch => ({
  onVote: ({ postId, topicId, value }) => dispatch(onVote({ postId, topicId, value })),
  onNotify: ({ topicId }) => dispatch(onNotify({ topicId })),
  onAccept: ({ topicId, postId }) => dispatch(onAccept({ topicId, postId }))
});

export default connect(
  null,
  mapDispatchToProps
)(PostHeaderMenu);
