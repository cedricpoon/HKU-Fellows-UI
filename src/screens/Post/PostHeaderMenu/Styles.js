import { StyleSheet } from 'react-native';
import { gridBase } from 'hkufui/theme/grid';
import { warningOrange, logo } from 'hkufui/theme/palette';

export default StyleSheet.create({
  separator: {
    height: 1,
    marginHorizontal: gridBase * 3
  },
  thumbUp: {
    color: warningOrange
  },
  thumbDown: {
    color: logo.red
  }
});
