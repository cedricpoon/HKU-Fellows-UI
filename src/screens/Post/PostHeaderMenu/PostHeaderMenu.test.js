import React from 'react';
import { View } from 'react-native';
import { shallow } from 'enzyme';
import { PostHeaderMenu } from './PostHeaderMenu';

describe('Testing PostHeaderMenu Component', () => {
  const requiredProps = {
    postId: 'postId',
    topicId: 'topicId',
    index: 1,
    position: {x: 0, y: 0},
    toggle: () => {}
  };

  it('renders as expected with PopupMenu required props', () => {
    const wrapper = shallow(
      <PostHeaderMenu  {...requiredProps}>
        <View></View>
      </PostHeaderMenu>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('has onRef', () => {
    const wrapper = shallow(
      <PostHeaderMenu  {...requiredProps} onRef={() => {}}>
        <View></View>
      </PostHeaderMenu>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('is solved', () => {
    const wrapper = shallow(
      <PostHeaderMenu {...requiredProps} solved>
        <View></View>
      </PostHeaderMenu>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
