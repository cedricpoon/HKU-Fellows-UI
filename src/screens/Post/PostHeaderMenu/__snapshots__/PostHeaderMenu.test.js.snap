// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing PostHeaderMenu Component has onRef 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PostHeaderMenu
    index={1}
    onRef={[Function]}
    position={
      Object {
        "x": 0,
        "y": 0,
      }
    }
    postId="postId"
    toggle={[Function]}
    topicId="topicId"
    uid="uid"
  >
    <View />
  </PostHeaderMenu>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        <Styled(Button)
          danger={true}
          iconLeft={true}
          onPress={[Function]}
          transparent={true}
        >
          <Styled(Icon)
            name="report"
            style={
              Object {
                "textAlign": "center",
                "width": 25,
              }
            }
            type="MaterialIcons"
          />
          <Styled(Text)>
            Report Abuse
          </Styled(Text)>
        </Styled(Button)>,
      ],
      "position": Object {
        "x": 0,
        "y": 0,
      },
      "postId": "postId",
      "toggle": [Function],
      "topicId": "topicId",
      "uid": "uid",
    },
    "ref": [Function],
    "rendered": Array [
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <Styled(Icon)
              name="report"
              style={
                Object {
                  "textAlign": "center",
                  "width": 25,
                }
              }
              type="MaterialIcons"
            />,
            <Styled(Text)>
              Report Abuse
            </Styled(Text)>,
          ],
          "danger": true,
          "iconLeft": true,
          "onPress": [Function],
          "transparent": true,
          "virtual": undefined,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "name": "report",
              "style": Object {
                "textAlign": "center",
                "width": 25,
              },
              "type": "MaterialIcons",
              "virtual": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Report Abuse",
              "virtual": undefined,
            },
            "ref": null,
            "rendered": "Report Abuse",
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          <Styled(Button)
            danger={true}
            iconLeft={true}
            onPress={[Function]}
            transparent={true}
          >
            <Styled(Icon)
              name="report"
              style={
                Object {
                  "textAlign": "center",
                  "width": 25,
                }
              }
              type="MaterialIcons"
            />
            <Styled(Text)>
              Report Abuse
            </Styled(Text)>
          </Styled(Button)>,
        ],
        "position": Object {
          "x": 0,
          "y": 0,
        },
        "postId": "postId",
        "toggle": [Function],
        "topicId": "topicId",
        "uid": "uid",
      },
      "ref": [Function],
      "rendered": Array [
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Styled(Icon)
                name="report"
                style={
                  Object {
                    "textAlign": "center",
                    "width": 25,
                  }
                }
                type="MaterialIcons"
              />,
              <Styled(Text)>
                Report Abuse
              </Styled(Text)>,
            ],
            "danger": true,
            "iconLeft": true,
            "onPress": [Function],
            "transparent": true,
            "virtual": undefined,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "name": "report",
                "style": Object {
                  "textAlign": "center",
                  "width": 25,
                },
                "type": "MaterialIcons",
                "virtual": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Report Abuse",
                "virtual": undefined,
              },
              "ref": null,
              "rendered": "Report Abuse",
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`Testing PostHeaderMenu Component is solved 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PostHeaderMenu
    index={1}
    position={
      Object {
        "x": 0,
        "y": 0,
      }
    }
    postId="postId"
    solved={true}
    toggle={[Function]}
    topicId="topicId"
    uid="uid"
  >
    <View />
  </PostHeaderMenu>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        <Styled(Button)
          danger={true}
          iconLeft={true}
          onPress={[Function]}
          transparent={true}
        >
          <Styled(Icon)
            name="report"
            style={
              Object {
                "textAlign": "center",
                "width": 25,
              }
            }
            type="MaterialIcons"
          />
          <Styled(Text)>
            Report Abuse
          </Styled(Text)>
        </Styled(Button)>,
      ],
      "position": Object {
        "x": 0,
        "y": 0,
      },
      "postId": "postId",
      "toggle": [Function],
      "topicId": "topicId",
      "uid": "uid",
    },
    "ref": null,
    "rendered": Array [
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <Styled(Icon)
              name="report"
              style={
                Object {
                  "textAlign": "center",
                  "width": 25,
                }
              }
              type="MaterialIcons"
            />,
            <Styled(Text)>
              Report Abuse
            </Styled(Text)>,
          ],
          "danger": true,
          "iconLeft": true,
          "onPress": [Function],
          "transparent": true,
          "virtual": undefined,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "name": "report",
              "style": Object {
                "textAlign": "center",
                "width": 25,
              },
              "type": "MaterialIcons",
              "virtual": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Report Abuse",
              "virtual": undefined,
            },
            "ref": null,
            "rendered": "Report Abuse",
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          <Styled(Button)
            danger={true}
            iconLeft={true}
            onPress={[Function]}
            transparent={true}
          >
            <Styled(Icon)
              name="report"
              style={
                Object {
                  "textAlign": "center",
                  "width": 25,
                }
              }
              type="MaterialIcons"
            />
            <Styled(Text)>
              Report Abuse
            </Styled(Text)>
          </Styled(Button)>,
        ],
        "position": Object {
          "x": 0,
          "y": 0,
        },
        "postId": "postId",
        "toggle": [Function],
        "topicId": "topicId",
        "uid": "uid",
      },
      "ref": null,
      "rendered": Array [
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Styled(Icon)
                name="report"
                style={
                  Object {
                    "textAlign": "center",
                    "width": 25,
                  }
                }
                type="MaterialIcons"
              />,
              <Styled(Text)>
                Report Abuse
              </Styled(Text)>,
            ],
            "danger": true,
            "iconLeft": true,
            "onPress": [Function],
            "transparent": true,
            "virtual": undefined,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "name": "report",
                "style": Object {
                  "textAlign": "center",
                  "width": 25,
                },
                "type": "MaterialIcons",
                "virtual": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Report Abuse",
                "virtual": undefined,
              },
              "ref": null,
              "rendered": "Report Abuse",
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`Testing PostHeaderMenu Component renders as expected with PopupMenu required props 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PostHeaderMenu
    index={1}
    position={
      Object {
        "x": 0,
        "y": 0,
      }
    }
    postId="postId"
    toggle={[Function]}
    topicId="topicId"
    uid="uid"
  >
    <View />
  </PostHeaderMenu>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        <Styled(Button)
          danger={true}
          iconLeft={true}
          onPress={[Function]}
          transparent={true}
        >
          <Styled(Icon)
            name="report"
            style={
              Object {
                "textAlign": "center",
                "width": 25,
              }
            }
            type="MaterialIcons"
          />
          <Styled(Text)>
            Report Abuse
          </Styled(Text)>
        </Styled(Button)>,
      ],
      "position": Object {
        "x": 0,
        "y": 0,
      },
      "postId": "postId",
      "toggle": [Function],
      "topicId": "topicId",
      "uid": "uid",
    },
    "ref": null,
    "rendered": Array [
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <Styled(Icon)
              name="report"
              style={
                Object {
                  "textAlign": "center",
                  "width": 25,
                }
              }
              type="MaterialIcons"
            />,
            <Styled(Text)>
              Report Abuse
            </Styled(Text)>,
          ],
          "danger": true,
          "iconLeft": true,
          "onPress": [Function],
          "transparent": true,
          "virtual": undefined,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "name": "report",
              "style": Object {
                "textAlign": "center",
                "width": 25,
              },
              "type": "MaterialIcons",
              "virtual": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Report Abuse",
              "virtual": undefined,
            },
            "ref": null,
            "rendered": "Report Abuse",
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          <Styled(Button)
            danger={true}
            iconLeft={true}
            onPress={[Function]}
            transparent={true}
          >
            <Styled(Icon)
              name="report"
              style={
                Object {
                  "textAlign": "center",
                  "width": 25,
                }
              }
              type="MaterialIcons"
            />
            <Styled(Text)>
              Report Abuse
            </Styled(Text)>
          </Styled(Button)>,
        ],
        "position": Object {
          "x": 0,
          "y": 0,
        },
        "postId": "postId",
        "toggle": [Function],
        "topicId": "topicId",
        "uid": "uid",
      },
      "ref": null,
      "rendered": Array [
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Styled(Icon)
                name="report"
                style={
                  Object {
                    "textAlign": "center",
                    "width": 25,
                  }
                }
                type="MaterialIcons"
              />,
              <Styled(Text)>
                Report Abuse
              </Styled(Text)>,
            ],
            "danger": true,
            "iconLeft": true,
            "onPress": [Function],
            "transparent": true,
            "virtual": undefined,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "name": "report",
                "style": Object {
                  "textAlign": "center",
                  "width": 25,
                },
                "type": "MaterialIcons",
                "virtual": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Report Abuse",
                "virtual": undefined,
              },
              "ref": null,
              "rendered": "Report Abuse",
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
